[3d] improve usability

gberaudo opened the pull request:
See https://jira.camptocamp.com/browse/GSLUX-18.

fgravin commented on geoportailv3/static/js/maincontroller.js:
```diff
@@ -442,8 +471,7 @@ app.MainController.prototype.createMap_ = function() {
       new ol.control.Zoom({zoomInLabel: '\ue032', zoomOutLabel: '\ue033'}),
-      new ol.control.ZoomToExtent({label: '\ue01b',
-        extent: this.defaultExtent_}),
+      new app.MainController.ZoomToExtent3d(this.defaultExtent_, () => this.ol3dm_),
```
Name is confusing as it's both for 2d and 3d.
Also the control generally depends on the cesiumManager extent.
I would postpone adding of this control after `old3dm` creation, and pass as argument `old3dm` `cameraExtentInRadians`.

gberaudo replied:
done

fgravin commented on geoportailv3/static/js/maincontroller.js:
```diff
@@ -428,6 +427,36 @@ app.MainController.prototype.addLocationControl_ =
     };
 
 
+app.MainController.ZoomToExtent3d = class extends ol.control.ZoomToExtent {
```
Not better in a separate file ?

gberaudo replied:
done

fgravin commented on geoportailv3/static/js/maincontroller.js:
```diff
@@ -507,15 +536,19 @@ app.MainController.Lux3DManager = class extends ngeo.olcs.Manager {
-  return new app.MainController.Lux3DManager(cesiumURL, this.map_, this.ngeoLocation_);
+  const cameraExtentInRadians = [5.31, 49.38, 6.64, 50.21].map(ol.math.toRadians);
+  const ol3dm = new app.MainController.Lux3DManager(cesiumURL, cameraExtentInRadians, this.map_, this.ngeoLocation_);
+  ol3dm.setRootScope($rootScope);
```
What is this used for ?

gberaudo replied:
You are right, it is weird. Looking at it and fixing what can be.

gberaudo replied:
done in the constructor and other refactorings.

gberaudo commented:
@fgravin, @tonio, the PR is now ready for a new review.
The PR:
- addresses all comments, including https://github.com/Geoportail-Luxembourg/geoportailv3/pull/1924#discussion_r153533637;
- fixed new toggle3d in minified build;
- moved 3d code in an "olcs" directory (olcs instead of 3d to be a valid js identifier);
- removed the unused 3d parameter;
- use spheres for limiting the extent;
- fixes the sinking under terrain issue.

tonio commented on geoportailv3/static/js/olcs/toggle3dcomponent.js:
```diff
@@ -5,22 +5,26 @@
 goog.require('ngeo.olcs.Service');
 
-app.toggle3d = {
+class Controller {
```
Any benefits to use a class based controller?

gberaudo replied:
No, it is simply more modern.
However you need to:
- have the constructor / class defined at the to level;
- have the `ngInject` annotation.

Otherwise the minified build will not work.

gberaudo commented:
Thanks for the review!

fgravin commented on geoportailv3/static/js/olcs/zoomtoextent.js:
```diff
@@ -0,0 +1,33 @@
+   */
+  handleZoomToExtent() {
+    if (this.ol3dm && this.ol3dm.is3dEnabled()) {
+      const cameraExtentInRadians = [5.31, 49.38, 6.64, 50.21].map(ol.math.toRadians);
```
My comment asked to get `cameraExtentInRadians` from `ol3dm` config.
`this.ol3dm.cameraExtentInRadians`
Here it's duplicated.

gberaudo replied:
OK, see https://github.com/Geoportail-Luxembourg/geoportailv3/pull/1933.
